### PR TITLE
KAFKA-13714: Fix cache flush position

### PR DESCRIPTION
**Pull request by vvcephei**
The caching store layers were passing down writes into lower store layers upon eviction, but not setting the context to the evicted records' context. Instead, the context was from whatever unrelated record was being processed at the time.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -207,7 +207,7 @@ if (file('.git').exists()) {
 } else {
   rat.enabled = false
 }
-println("Starting build with version $version (commit id ${commitId.take(8)}) using Gradle $gradleVersion, Java ${JavaVersion.current()} and Scala ${versions.scala}")
+println("Starting build with version $version (commit id ${commitId == null ? "null" : commitId.take(8)}) using Gradle $gradleVersion, Java ${JavaVersion.current()} and Scala ${versions.scala}")
```
Comment (vvcephei) [inline build.gradle:210]: For some reason, I was getting an NPE here while running the tests locally (once). I could revert this change, but it also seems harmless to keep it.

File: build.gradle [hunk 2]
```diff
@@ -435,6 +435,12 @@ subprojects {
       maxRetries = userMaxTestRetries
       maxFailures = userMaxTestRetryFailures
     }
+
+    // Allows devs to run tests in a loop to debug flaky tests
+    // Eg: I=0; while ./gradlew :streams:test -Prerun-tests --tests org.apache.kafka.streams.integration.IQv2StoreIntegrationTest --fail-fast; do (( I=$I+1 )); echo
+    if (project.hasProperty("rerun-tests")) {
+      outputs.upToDateWhen { false }
+    }
```
Comment (vvcephei) [inline build.gradle]: This is how I was able to repro the issue consistently, even though it only shows up roughly 1/500th of the time. I'll update the readme also to mention this approach instead of cleanTest, which causes us to recompile the tests on every iteration.

File: build.gradle [hunk 3]
```diff
@@ -468,6 +474,12 @@ subprojects {
...
```
Comment (vvcephei) [inline build.gradle]: Same thing for integration tests.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java [hunk 4]
```diff
@@ -226,10 +226,9 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
             if (rawNewValue != null || rawOldValue != null) {
                 // we need to get the old values if needed, and then put to store, and then flush
                 final ProcessorRecordContext current = context.recordContext();
-                context.setRecordContext(entry.entry().context());
-                wrapped().put(entry.key(), entry.newValue());
-
                 try {
+                    context.setRecordContext(entry.entry().context());
+                    wrapped().put(entry.key(), entry.newValue());
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:231]: we should be doing this inside the try block so that we will for sure set the context back no matter what happens.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java [hunk 5]
```diff
@@ -241,7 +240,13 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
                 }
             }
         } else {
-            wrapped().put(entry.key(), entry.newValue());
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:244]: This is why there was a bug under some configurations. We're leaving the wrong configuration in place while calling put on the wrapped store. We should set it to the evicted record's context before calling put. The same change is applied to the other two caching stores in this PR.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 6]
```diff
@@ -326,8 +326,8 @@ public synchronized void put(final Bytes key,
     public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
         try (final WriteBatch batch = new WriteBatch()) {
             dbAccessor.prepareBatch(entries, batch);
-            StoreQueryUtils.updatePosition(position, context);
             write(batch);
+            StoreQueryUtils.updatePosition(position, context);
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:330]: Minor fix. In all other places, we update the position after the write, not before.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java [hunk 7]
```diff
@@ -142,7 +142,9 @@ public static void updatePosition(
 
         if (stateStoreContext != null && stateStoreContext.recordMetadata().isPresent()) {
             final RecordMetadata meta = stateStoreContext.recordMetadata().get();
-            position.withComponent(meta.topic(), meta.partition(), meta.offset());
+            if (meta.topic() != null) {
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java:145]: Fixes some NPEs from the unit tests. I figure it's fine to have this here, since there's technically no contract that says topic can't be null and we're already guarding against a null context and missing recordMetadata.

File: streams/src/test/java/org/apache/kafka/streams/integration/IQv2StoreIntegrationTest.java [hunk 8]
```diff
@@ -760,45 +760,50 @@ public static void after() {
...
-                shouldHandleSessionKeyQueries();
+                if (storeToTest.isSession()) {
+                    // Note there's no "timestamped" differentiation here.
+                    // Idiosyncratically, SessionStores are _never_ timestamped.
+                    shouldHandleSessionKeyQueries();
+                }
             }
+        } catch (final AssertionError e) {
+            LOG.error("Failed assertion", e);
+            throw e;
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/IQv2StoreIntegrationTest.java:806]: Wrapping this and adding a log message helped me to correlate the actual moment of the test evaluation with the rest of the app logs in the test output. I figured I'd just leave it in and maybe help future generations.

File: streams/src/test/java/org/apache/kafka/streams/integration/IQv2StoreIntegrationTest.java [hunk 9]
```diff
@@ -1350,7 +1355,7 @@ private static Properties streamsConfiguration(final boolean cache, final boolea
                                                    final String supplier, final String kind) {
         final String safeTestName =
             IQv2StoreIntegrationTest.class.getName() + "-" + cache + "-" + log + "-" + supplier
-                + "-" + kind;
+                + "-" + kind + "-" + RANDOM.nextInt();
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/IQv2StoreIntegrationTest.java:1358]: Allows us to iterate quickly on the tests without colliding on the state stores or waiting for the session timeout.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingInMemoryKeyValueStoreTest.java [hunk 10]
```diff
@@ -221,17 +223,44 @@ public void shouldPutGetToFromCache() {
...
+    public void shouldMatchPositionAfterPutWithFlushListener() {
+        store.setFlushListener(record -> {}, false);
+        shouldMatchPositionAfterPut();
+    }
+
+    @Test
+    public void shouldMatchPositionAfterPutWithoutFlushListener() {
+        store.setFlushListener(null, false);
+        shouldMatchPositionAfterPut();
+    }
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingInMemoryKeyValueStoreTest.java]: If you look at the caching stores above, you'll see they have separate branches, depending on whether there's a flush listener defined, so we need to test both branches. Ditto with the other tests.

File: README.md [hunk 11]
```diff
@@ -37,13 +37,16 @@ Follow instructions in https://kafka.apache.org/quickstart
...
-    ./gradlew cleanTest integrationTest
+    ./gradlew -Prerun-tests test
+    ./gradlew -Prerun-tests unitTest
+    ./gradlew -Prerun-tests integrationTest
 
 ### Running a particular unit/integration test ###
     ./gradlew clients:test --tests RequestResponseTest
 
+### Repeatedly running a particular unit/integration test ###
+    I=0; while ./gradlew clients:test -Prerun-tests --tests RequestResponseTest --fail-fast; do (( I=$I+1 )); echo "Completed run: $I"; sleep 1; done
```
Comment (vvcephei) [inline README.md:48]: This is how I was able to reproduce the issue in only a few hours, even though it failed only about 1/500th of the time.

**Review by mjsax — APPROVED** [review]

**Comment by vvcephei** [comment]
Test failures unrelated:

```


Build / JDK 8 and Scala 2.12 / org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false | 0.38 sec | 1
-- | -- | --
Build / JDK 8 and Scala 2.12 / integration.kafka.server.FetchRequestBetweenDifferentIbpTest.testControllerNewIBP()

[Build / JDK 8 and Scala 2.12 / org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11926/5/testReport/org.apache.kafka.common.network/SslTransportLayerTest/Build___JDK_8_and_Scala_2_12____1__tlsProtocol_TLSv1_2__useInlinePem_false/)	0.38 sec	[1](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11926/5/)
 [Build / JDK 8 and Scala 2.12 / integration.kafka.server.FetchRequestBetweenDifferentIbpTest.testControllerNewIBP()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11926/5/testReport/integration.kafka.server/FetchRequestBetweenDifferentIbpTest/Build___JDK_8_and_Scala_2_12___testControllerNewIBP__/)
 ```
 
 The java 17 build is still queued after 5.5 hours.

**Comment by vvcephei** [comment]
Cherry-picked to 3.2

File: build.gradle (hunk 2 above)
Comment (chia7712) [reply build.gradle]: > which causes us to recompile the tests on every iteration.

just curious. on my local, `cleanTest` does not invoke recompile. The following console output is produced by command `./gradlew cleanTest clients:test --tests RequestResponseTest --info`
```
> Task :clients:compileTestJava UP-TO-DATE
Caching disabled for task ':clients:compileTestJava' because:
  Build cache is disabled
Skipping task ':clients:compileTestJava' as it is up-to-date.
Resolve mutations for :clients:testClasses (Thread[Execution worker Thread 6,5,main]) started.
:clients:testClasses (Thread[Execution worker Thread 6,5,main]) started.

> Task :clients:testClasses UP-TO-DATE
Skipping task ':clients:testClasses' as it has no actions.
Resolve mutations for :clients:checkstyleTest (Thread[Execution worker Thread 6,5,main]) started.
:clients:checkstyleTest (Thread[Execution worker Thread 6,5,main]) started.
```
Could you share the command which causes the recompile to me? thanks.
